syntax = "proto3";
package m2.pb;

import "m2g_AttributeType.proto";
import "m2g_ItemType.proto";
import "m2g_PropertyType.proto";
import "m2g_ResourceType.proto";
import "m2g_SpriteType.proto";
import "ProxyTypes.proto";

// A type that either carries an integer, or the real number type selected by the game
message IFF {
  oneof value_variant {
    int32 i = 1;
    FF ff = 2;
  }
}

message Property {
  m2g.pb.PropertyType type = 1;
  IFF iff = 2;
}

message Attribute2 {
  m2g.pb.AttributeType type = 1;
  IFF iff = 2;
}

// DEPRECATED
message Resource {
  m2g.pb.ResourceType type = 1;
  oneof value {
    float amount = 2;
  }
}

// DEPRECATED
message Attribute {
  m2g.pb.AttributeType type = 1;
  float amount = 2;
  oneof value_type {
    int32 int_value = 3;
    FF ff_value = 4;
  }
}

// DEPRECATED
enum Usage {
  // Item cannot be used, only held.
  STATIC = 0;

  // Item is constantly "in-use". The costs are deducted and benefits are applied automatically by the engine. When this
  // option is selected, the costs are benefits are pre-second values.
  AUTOMATIC = 1;

  // Item must be used manually and once at a time. The costs are deducted and benefits are applied by the engine. The
  // item doesn't disappear after use.
  REUSABLE = 2;

  // Item must be used manually and only once. The costs are deducted and benefits are applied by the engine. The item
  // disappears after one use.
  CONSUMABLE = 3;
}

/// An item also represents a capability
message Item {
  m2g.pb.ItemType type = 1;
  m2g.pb.ItemCategory category = 2;
  Usage usage = 3; // DEPRECATED
  bool use_on_acquire = 4; // DEPRECATED
  repeated Resource costs = 5; // DEPRECATED
  repeated Resource benefits = 6; // DEPRECATED
  repeated Resource acquire_benefits = 7; // Gained regardless use_on_acquire or not. Gained before benefits. DEPRECATED
  repeated Attribute attributes = 8; // DEPRECATED
  repeated Attribute2 attributes2 = 9;
  m2g.pb.SpriteType game_sprite = 10;
  m2g.pb.SpriteType ui_sprite = 11;
  string in_game_name = 12;
}

message Items {
  repeated Item items = 1;
}
